pkg/provider/terraform/instance/apply.go

```diff
@@ -18,30 +20,60 @@ func (p *plugin) terraformApply() error {
 	defer p.applyLock.Unlock()
 
 	if p.applying {
+		select {
+		case p.pollChannel <- true:
+			log.Infoln("Successfully interrupted terraform apply goroutine")
+		default:
+			log.Infoln("Polling channel is full, not interrupting")
+		}
 		return nil
 	}
 
+	p.pollChannel = make(chan bool, 1)
 	go func() {
+		initial := true
 		for {
-			if err := p.lock.TryLock(); err == nil {
-				defer p.lock.Unlock()
-				doTerraformApply(p.Dir)
-			} else {
-				log.Debugln("Can't acquire lock, waiting")
+			attempted, err := p.doTerraformApply(initial)
```

Comment:
I am not sure if you need to go through the trouble of adding channels to interrupt this loop just to prevent `doTerraformApply` from doing its thing if the plugin is not running on the leader node.   

There is a function that can locate the `Manager` object running on the same node.  You can simply call that to locate the Manager and then call `IsLeader()` and if it's not a leader, simply don't exec the `doTerraformApply`.  At the next time, this loop polls, same thing.  It's the same as having a separate polling goroutine but less complicated.

The helper is here:  https://github.com/docker/infrakit/blob/master/pkg/manager/discovery/discovery.go#L23

To call this, you do need a lookup function of the plugins... This is not currently plumbed through but the `infrakit` CLI offers an example.  I will add comments in your `main.go` so you know how to wire that.